Auto rollover

da-kami opened the pull request:
fixes #456 

Planning to write an actor test against it, but it won't be totally trivial so I am planning to do it in a follow up PR.

bonomat commented on daemon/src/taker_cfd.rs:
```diff
@@ -196,6 +198,26 @@ impl<O, M, W> Actor<O, M, W> {
     }
+
+    async fn handle_auto_rollover(&mut self) -> Result<()> {
+        tracing::debug!("Auto rollover");
```
Imho we should do a _smart_ auto rollover: 
check if the CFD's expiry is within 23h, if so, we rollover for another 24h. That way we can run it more regular for a low cost, e.g. every X minutes instead of hourly.

da-kami replied:
Note to future Daniel: `within 23h` will cover the CFD already being past attestation as well.

bonomat reviewed (commented):
> fixes #456
> 
> Todo:
> 
>     * [ ]  Change the rollover endpoint to toggle auto-rollover rather than allowing rollover itself.

I don't think we need the feature of toggling auto rollover.

rishflab committed:
Rollover CFD's automatically when a configurable rollover offset has elapsed

The rollover offset is the number of settlement interval after which rollover is triggered.

We want to be able to control the offset and settlement interval from the outside so we can change it in the tests.
A configurable settlement interval for the maker causes friction when implementing auto-rollover, because the settlement interval is not known to the taker.
The settlement interval should not be configurable, thus the configurable settlement interval is removed on the maker side.
The cfd actors and oracle actor use a global constant to decide the settlement interval and oracle even lookahead.
Not pulling the constant into the actors is a conscious decision, it allows us to set this to a different value in the actor tests.
The fact that we use the same constant for cfd actor and oracle actor is an implementation detail.

da-kami commented on daemon/src/lib.rs:
```diff
@@ -254,6 +256,12 @@ where
 
         tokio::spawn(oracle_ctx.run(oracle_constructor(cfds, Box::new(fan_out_actor))));
 
+        // TODO: Optimize initial auto-rollover to take time into account so that we don't want to
```
Outdated, should be removed

da-kami commented on daemon/src/taker.rs:
```diff
@@ -249,6 +250,7 @@ async fn main() -> Result<()> {
                 monitor::Actor::new(electrum, channel, cfds)
             }
         },
+        RollOverParams::new(1, time::Duration::hours(1))?,
```
This is the production code - why is the settlement interval set to `1 hour`?

da-kami commented on daemon/src/taker_cfd.rs:
```diff
@@ -167,6 +189,71 @@ impl<O, M, W> Actor<O, M, W> {
+            // trigger rollover if rollover window has elapsed
+            if cfd.expiry_timestamp() > now
+                && elapsed_since_creation
+                    > self.roll_over_params.rollover_offset
```
I cannot follow this logic, please explain what the values represent after your changes and how they relate to each other.

da-kami reviewed (commented):
@rishflab it would be great if you could open a new PR, so I can properly give this a review. Since this PR was created by me I can only comment - but the code does not reflect the initial solution anymore. 

I am confused by how the solution was changed.

Please explain in more detail:

* Why is the settlement interval set to `1 hour` in production code? It was set to `24 hours` before with an offset of `1 hour`. The settlement interval represents when the CFD will be settled (it defines how far in the future the oracle event should be). With the current changes that is not reflected anymore.
* The commit message does not reflect the code at all, it is my original commit message! As far as I can see e.g. the changes for removing the configurable settlement interval on the maker were dropped - why?

da-kami commented on daemon/tests/happy_path.rs:
```diff
@@ -59,6 +66,167 @@ async fn taker_takes_order_and_maker_rejects() {
+    assert!(matches!(taker_cfd.state, CfdState::PendingOpen { .. }));
+    assert!(matches!(maker_cfd.state, CfdState::PendingOpen { .. }));
+
+    taker.trigger_auto_rollover().await;
```
why is this called here already and then again later in the test?

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -680,20 +680,26 @@ impl Cfd {
     /// extreme case where a user publishes the commit transaction right after the contract was
     /// initialized. In this case, the oracle still has `1.0 *
-    /// cfdorder.settlement_time_interval_hours` time to attest and no one can publish the refund
+    /// cfdorder.settlement_interval` time to attest and no one can publish the refund
```
Can you put this rename into a different commit?

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -82,6 +101,7 @@ where
         + xtra::Handler<wallet::Sign>
         + xtra::Handler<wallet::BuildPartyParams>,
 {
+    #[allow(clippy::too_many_arguments)]
```
That seems pretty useless.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -60,6 +60,24 @@ enum RollOverState {
+
+impl RollOverParams {
+    pub fn new(rollover_offset: u32, settlement_interval: Duration) -> Result<Self> {
+        if rollover_offset < 1 {
+            bail!("It does not make sense to rollover a cfd until at least 1 settlement interval has passed")
+        }
```
Use `NonZeroU32` instead.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -167,6 +189,71 @@ impl<O, M, W> Actor<O, M, W> {
+                    > self.roll_over_params.rollover_offset
+                        * self.roll_over_params.settlement_interval
+            {
+                tracing::debug!("proposed roll over");
```
Make this log part of the `propose_roll_over` function and move it to the bottom so we don't erroneously log this in case the function fails.

thomaseizinger commented on daemon/tests/happy_path.rs:
```diff
@@ -59,6 +66,167 @@ async fn taker_takes_order_and_maker_rejects() {
+
+    #[allow(clippy::redundant_closure)] // clippy is in the wrong here
+    taker
+        .mocks
+        .wallet()
+        .await
+        .expect_build_party_params()
+        .returning(|msg| build_party_params(msg));
+
+    #[allow(clippy::redundant_closure)] // clippy is in the wrong here
+    maker
+        .mocks
+        .oracle()
+        .await
+        .expect_monitor_attestation()
+        .return_const(());
+
+    #[allow(clippy::redundant_closure)] // clippy is in the wrong here
+    taker
+        .mocks
+        .oracle()
+        .await
+        .expect_monitor_attestation()
+        .return_const(());
+
+    #[allow(clippy::redundant_closure)] // clippy is in the wrong here
+    maker
+        .mocks
+        .monitor()
+        .await
+        .expect_start_monitoring()
+        .return_const(());
+
+    #[allow(clippy::redundant_closure)] // clippy is in the wrong here
+    taker
+        .mocks
+        .monitor()
+        .await
+        .expect_start_monitoring()
+        .return_const(());
```
Please reduce this noise similar to what @klochowicz did in https://github.com/itchysats/itchysats/pull/511.

thomaseizinger commented on daemon/tests/happy_path.rs:
```diff
@@ -59,6 +66,167 @@ async fn taker_takes_order_and_maker_rejects() {
+        .expect_start_monitoring()
+        .return_const(());
+
+    #[allow(clippy::redundant_closure)] // clippy is in the wrong here
```
There is no closure in this statement?

thomaseizinger commented on daemon/tests/happy_path.rs:
```diff
@@ -59,6 +66,167 @@ async fn taker_takes_order_and_maker_rejects() {
+    );
+}
+
+// Helper function setting up a "happy path" wallet mock
+fn mock_wallet_sign_and_broadcast(wallet: &mut MutexGuard<'_, harness::mocks::wallet::MockWallet>) {
+    let mut seq = mockall::Sequence::new();
+    wallet
+        .expect_sign()
+        .times(1)
+        .returning(|sign| Ok(sign.psbt))
+        .in_sequence(&mut seq);
+    wallet
+        .expect_broadcast()
+        .times(1)
+        .returning(|_| Ok(dummy_tx_id()))
+        .in_sequence(&mut seq);
+}
+
```
We have moved on to a different pattern, see https://github.com/itchysats/itchysats/pull/511.

thomaseizinger commented on daemon/tests/harness/mod.rs:
```diff
@@ -151,6 +172,8 @@ impl Taker {
         oracle_pk: schnorrsig::PublicKey,
         maker_address: SocketAddr,
         maker_noise_pub_key: x25519_dalek::PublicKey,
+        settlement_interval: Duration,
+        rollover_offset: u32,
```
Do we ever configure this with something else other than the hardcoded values further up? If not, I'd prefer to not (yet) expose this as parameters.

thomaseizinger commented on daemon/tests/happy_path.rs:
```diff
@@ -59,6 +66,167 @@ async fn taker_takes_order_and_maker_rejects() {
+
+    taker.trigger_auto_rollover().await;
+
+    loop {
+        let update_proposals = next(&mut maker.update_proposals).await;
+        if let Some(update_proposal) = update_proposals.get(&maker_cfd.order.id) {
+            assert!(matches!(
+                update_proposal,
+                UpdateCfdProposal::RollOverProposal { .. }
+            ));
+            break;
+        }
+    }
+
```
Extract this into a helper please so we don't clutter up the test.

thomaseizinger replied:
The helper should return the next update proposal by ID. The assertion should stay in the test.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -167,6 +189,71 @@ impl<O, M, W> Actor<O, M, W> {
+    }
+
+    async fn propose_roll_over(&mut self, order_id: OrderId) -> Result<()> {
+        if self.current_pending_proposals.contains_key(&order_id) {
+            anyhow::bail!("An update for order id {} is already in progress", order_id)
+        }
```
Can this ever be hit? We cleared the list on line 200?

rishflab replied:
not currently, but if someones makes `propose_rollover` pub in the future, it would be nice to have this check. I dont think it does any harm to keep it?

thomaseizinger replied:
It does harm. It introduces complexity because its presence makes you think that it can be hit. Unnecessary code is one of the worst forms of complexity because you end up wasting hours of looking at it to only realize that is unrelated to the problem you are debugging.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -680,20 +680,26 @@ impl Cfd {
     }
 
-    /// A factor to be added to the CFD order settlement_time_interval_hours for calculating the
+    pub fn creation_timestamp(&self) -> Result<OffsetDateTime> {
+        let timestamp =
+            OffsetDateTime::from_unix_timestamp(self.order.creation_timestamp.seconds())?;
+        Ok(timestamp)
+    }
```
I think we have some unnecessary conversions here. Can we get rid of these failure paths or at least change them to panics because they are unreachable?

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -60,6 +60,24 @@ enum RollOverState {
 
+pub struct RollOverParams {
+    settlement_interval: Duration,
+    /// The number of settlement intervals in which an auto-rollover is triggered
+    rollover_offset: u32,
```
Naming and docs need improvement. It is hard to workout what this value does without reading the code.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -167,6 +189,71 @@ impl<O, M, W> Actor<O, M, W> {
+            .filter(|cfd| matches!(cfd.state, CfdState::Open { .. }))
+        {
+            tracing::debug!(
+                "settlement_interval: {:?}",
```
Debug formatting a duration is going to be noisy. Please either do seconds or something like this: https://docs.rs/humantime/2.1.0/humantime/fn.format_duration.html

rishflab replied:
removing, i don't think we need to print this anymore
